src/chain/chain_accumulator.rs

```diff
@@ -9,38 +9,54 @@
 use super::{AccumulatingEvent, NetworkEvent, Proof, ProofSet, SectionInfoSigPayload};
 use crate::id::PublicId;
 use log::LogLevel;
-use std::collections::{BTreeMap, BTreeSet, VecDeque};
-use std::mem;
+use std::collections::{BTreeMap, BTreeSet};
+use std::{cmp, mem};
 
 /// The threshold (number of unvoted votes) a node to be considered as unresponsive.
-pub const UNRESPONSIVE_THRESHOLD: usize = 48;
+pub const UNRESPONSIVE_THRESHOLD: usize = 96;
 /// The period (X consensued observations) after which node be considered as unresponsive.
-pub const UNRESPONSIVE_WINDOW: usize = 64;
+pub const UNRESPONSIVE_WINDOW: usize = 128;
+/// The max number of peers that under the responsiveness tracking.
+pub const UNRESPONSIVE_PEERS: usize = 16;
 
-struct VoteStatus {
-    event: AccumulatingEvent,
-    unvoted: BTreeSet<PublicId>,
+#[derive(Default)]
+struct VoteStatuses {
+    unvoted: BTreeMap<PublicId, usize>,
 }
 
-impl VoteStatus {
-    fn new(event: AccumulatingEvent, all_voters: BTreeSet<PublicId>) -> Self {
-        VoteStatus {
-            event: event,
-            unvoted: all_voters,
+impl VoteStatuses {
+    fn add_expectation(
+        &mut self,
+        all_voters: BTreeSet<PublicId>,
+        all_members: &BTreeSet<PublicId>,
+    ) {
+        for id in all_voters {
+            let count = self.unvoted.entry(id).or_insert_with(|| 0);
+            // Avoid punishing for a missed vote long time ago.
+            *count = cmp::min(UNRESPONSIVE_WINDOW, count.saturating_add(1));
+        }
+
+        // Avoid overflowing and pruning too often.
+        if self.unvoted.len() > UNRESPONSIVE_PEERS {
```

Comment:
Not sure about that.

Reply:
I think idealy the UNRESPONSIVE_PEERS shall be say 4 times of the elder_size.

Reply:
Not sure about that either.
It seem it could be a simple sanity check to avoid going through the list all the time, but only when a voter in unvoted is no longer in all_members `self.unvoted.len() > all_members.len()`.